### PR TITLE
fix doxygen warnings in cudf/io/types.hpp, other header files

### PR DESCRIPTION
**Pull request by karthikeyann**
Fixes parts of https://github.com/rapidsai/cudf/issues/9373
added missing documentation to fix doxygen warnings in multiple files

fixes 93 warnings.

### CODE REVIEW

File: cpp/include/cudf/io/datasource.hpp [hunk 1]
```diff
@@ -64,6 +68,13 @@ class datasource {
      */
     virtual ~buffer() {}
 
+    /**
+     * @brief Factory to construct a datasource object from a container.
+     *
+     * @tparam Container Type of the container to construct the datasource from
+     * @param data_owner The container to construct the datasource from (ownership is transferred)
+     * @return Constructed datasource object
```
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:76]:
```suggestion
    /**
     * @brief Factory to construct a datasource buffer object from a container.
     *
     * @tparam Container Type of the container to construct the buffer from
     * @param data_owner The container to construct the buffer from (ownership is transferred)
     * @return Constructed buffer object
```

File: cpp/include/cudf/io/datasource.hpp [hunk 2]
```diff
@@ -285,6 +317,8 @@ class datasource {
    public:
     /**
      * @brief Moves the input container into the newly created object.
+     *
+     * @param data_owner The container to move
```
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:321]:
```suggestion
     * @param data_owner The container to construct the buffer from (ownership is transferred)
```

File: cpp/include/cudf/io/datasource.hpp [hunk 3]
```diff
@@ -294,14 +328,28 @@ class datasource {
     /**
      * @brief Moves the input container into the newly created object, and exposes a subspan of the
      * buffer.
+     *
+     * @param data_owner The container to move
```
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:332]:
```suggestion
     * @param data_owner The container to construct the buffer from (ownership is transferred)
```

File: cpp/include/cudf/io/datasource.hpp [hunk 4]
```diff
@@ -314,6 +362,13 @@ class datasource {
   };
 };
 
+/**
+ * @brief Factory to construct a datasource object from a container.
+ *
+ * @tparam Container Type of the container to construct the datasource from
+ * @param data_owner The container to construct the datasource from (ownership is transferred)
+ * @return Constructed datasource object
```
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:370]:
```suggestion
 * @brief Factory to construct a datasource buffer object from a container.
 *
 * @tparam Container Type of the container to construct the buffer from
 * @param data_owner The container to construct the buffer from (ownership is transferred)
 * @return Constructed buffer object
```

File: cpp/include/cudf/io/datasource.hpp [hunk 5]
```diff
@@ -294,14 +328,28 @@ class datasource {
...
+     * @brief Returns the size of the buffer.
+     *
+     * @return The size of the buffer in bytes
+     */
     [[nodiscard]] size_t size() const override { return _size; }
 
+    /**
+     * @brief Returns the pointer to the buffer.
+     *
+     * @return Pointer to the buffer
```
Comment (vuule) [inline cpp/include/cudf/io/datasource.hpp:351]:
```suggestion
     * @brief Returns the pointer to the data in the buffer.
     *
     * @return Pointer to the data in the buffer
```

**Review by vuule — CHANGES_REQUESTED** [review]
Looks good for the most part. A few corrections/questions:

File: cpp/include/cudf/io/types.hpp [hunk 6]
```diff
@@ -138,8 +143,8 @@ struct table_metadata {
  * @brief Table with table metadata used by io readers to return the metadata by value
  */
 struct table_with_metadata {
-  std::unique_ptr<table> tbl;
-  table_metadata metadata;
+  std::unique_ptr<table> tbl;  //!< Table
+  table_metadata metadata;     //!< Table metadata
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/io/types.hpp:147]: This file had similar style at other places, so use this `//!<` for uniformity. No other reason.

File: cpp/include/cudf/io/datasource.hpp (hunk 4 above)
Comment (karthikeyann) [reply cpp/include/cudf/io/datasource.hpp:370]: This docstring is copied from the declaration part of the code in same file. doxygen throws warning due to a limitation of template function.
shall we move the code definition to the declaration itself?

**Comment by ajschmidt8** [comment]
Removing `ops-codeowners` from the required reviews since it doesn't seem there are any file changes that we're responsible for. Feel free to add us back if necessary.

File: cpp/include/cudf/io/datasource.hpp (hunk 4 above)
Comment (vuule) [reply cpp/include/cudf/io/datasource.hpp]: You can move the definition to the declaration, just need to forward declare `owning_buffer`. I think it's also possible to `copydoc` and avoid doc duplication that way.

Comment (karthikeyann) [reply cpp/include/cudf/io/datasource.hpp]: I tried `copydoc`, doxygen complains recursive reference.

Comment (karthikeyann) [reply cpp/include/cudf/io/datasource.hpp]: done.

**Review by vuule — APPROVED** [review]

File: cpp/include/cudf/io/types.hpp [hunk 7]
```diff
@@ -362,61 +505,82 @@ class column_in_metadata {
...
   /**
    * @brief Get the number of children of this column
+   *
+   * @return The number of children of this column
    */
   [[nodiscard]] size_type num_children() const { return children.size(); }
 };
 
+/**
+ * @brief Meta data for a table
```
Comment (vyasr) [inline cpp/include/cudf/io/types.hpp:582]:
```suggestion
 * @brief Metadata for a table
```

**Review by vyasr — APPROVED** [review]
One minor typo fix. Also, I asked this on another of your PRs: I'm not sure why `pure` is needed based on the contents of the doxygen docs for the pure command. I'm approving assuming you can address that question consistently across all the doxygen PRs. Thanks!

**Comment by karthikeyann** [comment]
@gpucibot merge

**Pull request merged**